File: Preg1/fibonacci3-recursivo2.hs
import System.IO
import Data.Array

fibo    :: Integer -> Array Integer Integer
fibo n  =  a  where a = array (0,n) ([(0, 0), (1, 1), (2,1)] ++ 
                                     [(i,  a!(i-3) + a!(i-2) + a!(i-1)) | i <- [3..n]])
main = do
    putStrLn "¿Cuantos elementos?: "
    x <- getLine
    let n = read x :: Integer
    print(fibo (n-1))

File: Preg1/fibonacci3-recursivo2.scala
object Fibonacci {
    def fibo3(a:Int,b:Int,c:Int,x:Int,n:Int) : Unit =  {
        if(x<=n) then {
            x match
                case 1 => print ((a)+ " ")
                case 2 => print ((b)+ " ")
                case 3 => print ((c)+ " ")
                case _ => print ((a+b+c)+ " ")
            if (x>3) {
                fibo3(b,c,a+b+c,x+1,n)
            } else {
                fibo3(a,b,c,x+1,n)
            }
        }
    }
    
    def fibo(n:Int) : Unit  =  {
       fibo3(0,1,1,1,n)
    }
    def main(args: Array[String]) = {
        println ("¿Cuantos elementos?: ")
        val x = scala.io.StdIn.readInt()
        fibo(x)
    }
} 